flink-connectors/flink-connector-kafka-0.11/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011.java

```diff
@@ -0,0 +1,1000 @@
+			String defaultTopicId,
+			KeyedSerializationSchema<IN> serializationSchema,
+			Properties producerConfig,
+			FlinkKafkaPartitioner<IN> customPartitioner,
+			Semantic semantic,
+			int kafkaProducersPoolSize) {
+		super(
+			TypeInformation.of(KafkaTransactionState.class),
+			TypeInformation.of(new TypeHint<List<KafkaTransactionState>>() {}));
+
+		requireNonNull(defaultTopicId, "TopicID not set");
+		requireNonNull(serializationSchema, "serializationSchema not set");
+		requireNonNull(producerConfig, "producerConfig not set");
+		ClosureCleaner.clean(customPartitioner, true);
+		ClosureCleaner.ensureSerializable(serializationSchema);
+
+		this.defaultTopicId = defaultTopicId;
+		this.schema = serializationSchema;
+		this.producerConfig = producerConfig;
+		this.flinkKafkaPartitioner = customPartitioner;
+		this.semantic = semantic;
+		this.kafkaProducersPoolSize = kafkaProducersPoolSize;
+
+		// set the producer configuration properties for kafka record key value serializers.
+		if (!producerConfig.containsKey(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG)) {
+			this.producerConfig.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class.getCanonicalName());
+		} else {
+			LOG.warn("Overwriting the '{}' is not recommended", ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG);
+		}
+
+		if (!producerConfig.containsKey(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG)) {
+			this.producerConfig.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class.getCanonicalName());
+		} else {
+			LOG.warn("Overwriting the '{}' is not recommended", ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG);
+		}
+
+		// eagerly ensure that bootstrap servers are set.
+		if (!this.producerConfig.containsKey(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG)) {
+			throw new IllegalArgumentException(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG + " must be supplied in the producer config properties.");
+		}
+
+		this.topicPartitionsMap = new HashMap<>();
+	}
+
+	// ---------------------------------- Properties --------------------------
+
+	/**
+	 * Defines whether the producer should fail on errors, or only log them.
+	 * If this is set to true, then exceptions will be only logged, if set to false,
+	 * exceptions will be eventually thrown and cause the streaming program to
+	 * fail (and enter recovery).
+	 *
+	 * <p>Method is only accessible for approach (a) (see above)
+	 *
+	 * @param logFailuresOnly The flag to indicate logging-only on exceptions.
+	 */
+	public void setLogFailuresOnly(boolean logFailuresOnly) {
+		this.logFailuresOnly = logFailuresOnly;
+	}
+
+	// ----------------------------------- Utilities --------------------------
+
+	/**
+	 * Initializes the connection to Kafka.
+	 *
+	 * <p>This method is used for approach (a) (see above).
+	 */
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		if (semantic != Semantic.NONE && !((StreamingRuntimeContext) this.getRuntimeContext()).isCheckpointingEnabled()) {
```

Comment:
I'm confused. In some previous offline discussion you (or was it maybe Stephan?) told me that we need separate enum for setting up `FlinkKafaProducer011`. What's the point of having this separate enum, if we want to tie it up to `CheckpointMode`?